lib/internal/modules/esm/loader.js

```diff
@@ -163,72 +179,83 @@ class Loader {
     return module.getNamespace();
   }
 
+  async importLoader(specifier) {
+    if (cwd === undefined) {
+      try {
+        // `process.cwd()` can fail.
+        cwd = process.cwd() + '/';
+      } catch {
+        cwd = 'file:///';
+      }
+      cwd = pathToFileURL(cwd).href;
+    }
+
+    const { url } = await defaultResolve(specifier, cwd,
+                                         { conditions: DEFAULT_CONDITIONS });
+    const { format } = await defaultGetFormat(url, {});
+
+    // !!! CRITICAL SECTION !!!
+    // NO AWAIT OPS BETWEEN HERE AND SETTING JOB IN MODULE MAP
+    let job = this.moduleMap.get(url);
+    if (job === undefined) {
+      if (!translators.has(format))
+        throw new ERR_UNKNOWN_MODULE_FORMAT(format);
+
+      const loaderInstance = translators.get(format);
+
+      job = new ModuleJob(this, url, loaderInstance, false, false);
+      this.moduleMap.set(url, job);
+      // !!! END CRITICAL SECTION !!!
+    }
+
+    const { module } = await job.run();
+    return module.getNamespace();
+  }
+
   hook(hooks) {
     const {
       resolve,
-      dynamicInstantiate,
       getFormat,
       getSource,
-      transformSource,
-      getGlobalPreloadCode,
```

Comment:
So I guess this PR also removes `dynamicInstantiate`, which is fine; are the others removed from this list because they behave in idiosyncratic ways (`transformSource` chains in reverse, `getGlobalPreloadCode` runs sequentially but unchained?).